pyccel/ast/operators.py

```diff
@@ -0,0 +1,949 @@
+
+        If the arguments are numeric then the dtype and precision
+        match the broadest type and the largest precision
+        e.g.
+            1 + 2j -> PyccelAdd(LiteralInteger, LiteralComplex) -> complex
+        """
+        integers  = [a for a in self._args if a.dtype in (NativeInteger(),NativeBool())]
+        reals     = [a for a in self._args if a.dtype is NativeReal()]
+        complexes = [a for a in self._args if a.dtype is NativeComplex()]
+        strs      = [a for a in self._args if a.dtype is NativeString()]
+
+        if strs:
+            self._handle_str_type(strs)
+            assert len(integers + reals + complexes) == 0
+        elif complexes:
+            self._handle_complex_type(complexes)
+        elif reals:
+            self._handle_real_type(reals)
+        elif integers:
+            self._handle_integer_type(integers)
+        else:
+            raise TypeError('cannot determine the type of {}'.format(self))
+
+
+    @staticmethod
+    def _handle_str_type(strs):
+        """
+        Set dtype and precision when both arguments are strings
+        """
+        raise TypeError("unsupported operand type(s) for /: 'str' and 'str'")
+
+    def _handle_complex_type(self, complexes):
+        """
+        Set dtype and precision when the result is complex
+        """
+        self._dtype     = NativeComplex()
+        self._precision = max(a.precision for a in complexes)
+
+    def _handle_real_type(self, reals):
+        """
+        Set dtype and precision when the result is real
+        """
+        self._dtype     = NativeReal()
+        self._precision = max(a.precision for a in reals)
+
+    def _handle_integer_type(self, integers):
+        """
+        Set dtype and precision when the result is integer
+        """
+        self._dtype     = NativeInteger()
+        self._precision = max(a.precision for a in integers)
+
+    def _set_shape_rank(self):
+        """ Sets the shape and rank
+
+        Strings must be scalars.
+
+        For numeric types the rank and shape is determined according
+        to numpy broadcasting rules where possible
+        """
+        if self._dtype is NativeString():
+            self._rank  = 0
+            self._shape = ()
+        else:
+            ranks  = [a.rank for a in  self._args]
+            shapes = [a.shape for a in self._args]
+
+            if None in ranks:
+                self._rank  = None
+                self._shape = None
+
+            elif all(sh is not None for tup in shapes for sh in tup):
+                shape = broadcast(self._args[0].shape, self._args[1].shape)
+
+                self._shape = shape
+                self._rank  = len(shape)
+            else:
+                self._rank  = max(a.rank for a in self._args)
+                self._shape = [None]*self._rank
+
+#==============================================================================
+
+class PyccelArithmeticOperator(PyccelBinaryOperator):
```

Comment:
Is anything a `PyccelBinaryOperator` but not a `PyccelArithmeticOperator`?

Reply:
Yes. Comparisons

Reply:
There is a `PyccelComparisonOperator`...